source/optouts/OptOutsService.cs

```
@@ -0,0 +1,157 @@
+        {
+            return GetByPhoneNumber(null, accountReference, pageNumber, pageSize);
+        }
+
+        public SubscriptionCollection GetAll(int pageNumber, int pageSize)
+        {
+            return GetByPhoneNumber(null, null, pageNumber, pageSize);
+        }
+
+        private int GetStartIndex(int pageNumber, int pageSize)
+        {
+            if (pageNumber < 1) throw new ArgumentException("Page number must be greater than zero.", "pageNumber");
+            if (pageSize < 1) throw new ArgumentException("Page size must be greater than zero.", "pageSize");
+
+            var startIndex = ((--pageNumber)*pageSize);
+            return startIndex;
+        }
+
+        public OptOut Add(string accountReference, string phoneNumber)
+        {
+            var requestData = new OptOutCreateRequest
+            {
+                AccountReference = accountReference,
+                From = new FromAddress
+                {
+                    PhoneNumber = phoneNumber
+                }
+            };
+
+            var request = Request.Create("POST", new Uri(string.Concat(_baseUrl, "v1.0/optouts")))
+                                 .WithHeader("Authorization", "Basic " + _credentials.EncodedValue())
+                                 .WithAcceptHeader(Constants.XML_MEDIA_TYPE)
+                                 .If(_credentials.UseProxy, r => r.WithProxy(_credentials.WebProxy))
+                                 .WriteBody(Constants.XML_MEDIA_TYPE, streamWriter => streamWriter.Write(new XmlSerialiser().Serialise(requestData)));
+
+            try
+            {
+                var response = request.GetResponse();
+                using (var reader = new StreamReader(response.GetResponseStream(), Encoding.UTF8))
+                {
+                    string responseText = reader.ReadToEnd();
+                    return new XmlSerialiser().Deserialise<OptOutCreateRootResponse>(responseText).OptOut;
+                    

```

this swallows the structured errors?